### DISCUSSION TITLE
Macro discussion thread

### DISCUSSION THREAD

**Discussion by migueldeicaza** [migueldeicaza/SwiftGodot#94] [General]
To improve the usability of SwiftGodot, it would be nice to leverage the recently released macros.

Some good ideas are in this PR: 

https://github.com/migueldeicaza/SwiftGodot/pull/89

And in this bug (which links to a POC):

https://github.com/migueldeicaza/SwiftGodot/issues/86

**Comment by alicerunsonfedora** [comment]
Pertaining to https://github.com/migueldeicaza/SwiftGodot/pull/89#issuecomment-1586254365:

This was my line of thinking when it came to the `@Export` macro. While having the hints as parameters and whatnot makes sense, I do feel it does start to pollute the parameters of `@Export`, and it still adds ambiguity to the member variable in question.

My initial line of thinking was to have macros such as:

```swift
@ExportedRange(in: 1...100, stride: 1) var bodyCount = 1
@ExportedCheckbox var debugSettings = false
@ExportedFile(allowing: [.gdscript, .jenson]) var resourcePath = "res://"
@ExportedField(style: .single) var someText = ""
```

While it does mean that there are more macros in place, I see a couple of advantages this way:
- First, it's immediately clear what the export is doing. One of the goals for Swift macros is that it shouldn't be "magic", and I think this clarity helps in ensuring this.
  - On this note, it also helps disambiguate what a certain variable type is. For example, if I have an `@Export var something: String = ""`, it's not completely clear what this does. Is it a file path? A text field? Whereas specifying `@ExportedFile(allowing: [.json)` makes it clear that the variable is a file path limited to JSON files.
- Also, it allows for more expressive constructions. Rather than trying to make a generic "one size fits all" approach, this would allow for each macro to have finely tuned arguments to it.

**Reply by alicerunsonfedora** [reply]
On top of this, it seems that exports work similarly in GDScript 4 for types such as files, categories, etc.: https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_exports.html.

**Comment by migueldeicaza** [comment]
I like the way you are thinking about this, perhaps Export can be just for methods, but for properties that have a richer way of being surfaced in the IDE, it might make sense to have additional values.   

Nice find on the GDScript export attributes!
